source/merge-deep.d.ts

```diff
@@ -0,0 +1,412 @@
+	[Key in keyof Destination | keyof Source]: Key extends keyof Source
+		? // Source found, check for destination
+			Key extends keyof Destination
+			? // Both source and destination exists
+				Options['recordMergeMode'] extends 'merge-or-replace' // Default value first
+				? MergeDeepOrReturn<Source[Key], Destination[Key], Source[Key], Options>
+				: Options['recordMergeMode'] extends 'merge-or-union'
+				? MergeDeepOrReturn<Destination[Key] | Source[Key], Destination[Key], Source[Key], Options>
+				: Options['recordMergeMode'] extends 'union'
+				? Destination[Key] | Source[Key]
+				: // Assume recordMergeMode = 'replace'
+					Source[Key]
+			: // Only the source exists
+				Source[Key]
+		: // No source, take the destination
+		Key extends keyof Destination
+		? Destination[Key]
+		: never; // (this test is useless, but make TS happy, It can never be never)
+};
+
+/**
+Strips `undefined` values on first and second levels.
+*/
+type StripUndefinedValues<Type> = ConditionalExcept<{
+	[Key in keyof Type ]: Type[Key] extends UnknownRecord ? ConditionalSimplify<StripUndefinedValues<Type[Key]>> : Type[Key];
+}, undefined>;
+
+/**
+Wrapper around {@link DoMergeRecord} which defines whether or not to strip `undefined` values.
+*/
+type MergeDeepRecord<
+	Destination,
+	Source,
+	Options extends MergeDeepOptions,
+> = Options['stripUndefinedValues'] extends true
+	? StripUndefinedValues<DoMergeRecord<Destination, Source, Options>>
+	: DoMergeRecord<Destination, Source, Options>;
+
+/**
+Try to merge two objects or two arrays/tuples recursively into a new type or return the default value.
+*/
+type MergeDeepOrReturn<DefaultValue, Destination, Source, Options> = Destination extends UnknownArrayOrTuple
+	? Source extends UnknownArrayOrTuple
+		? ConditionalSimplify<MergeDeepArray<Destination, Source, Merge<MergeDeepDefaultOptions, Options>>, Function>
+		: DefaultValue
+	: Destination extends UnknownRecord
+	? Source extends UnknownRecord
+		? ConditionalSimplify<MergeDeepRecord<Destination, Source, Merge<MergeDeepDefaultOptions, Options>>, Function>
+		: DefaultValue
+	: DefaultValue;
+
+// Merge mode for object properties.
+type MergeDeepRecordMode = 'merge-or-replace' | 'merge-or-union' | 'replace' | 'union';
+
+// Merge mode for array/tuple elements.
+type MergeDeepArrayMode = MergeDeepRecordMode | 'merge-or-spread' | 'spread';
+
+/**
+MergeDeep options.
+
+@see MergeDeep
+*/
+export interface MergeDeepOptions {
+	/**
+	Should strip `undefined` values from the resulting type.
+
+	@default false
+	*/
+	stripUndefinedValues?: boolean;
+
+	/**
+	Merge mode for object properties.
+
+	When we walk through the properties of the objects and the same key is found in both, a merge mode must be chosen:
+	- `merge-or-replace`: Try to merge the values for that key, otherwise replace the destination value by the source value.
+	- `merge-or-union`: Try to merge the values for that key, otherwise union the destination value and the source value.
```

Comment:
As a user, it's hard for me to see in what situations I would want which options.